core/src/main/java/org/jruby/RubyKernel.java

```diff
@@ -163,66 +164,46 @@ static void recacheBuiltinMethods(Ruby runtime) {
-    public static IRubyObject autoload(ThreadContext context, final IRubyObject recv, IRubyObject symbol, IRubyObject file) {
-        final Ruby runtime = context.runtime;
-        final String nonInternedName = symbol.asJavaString();
+        final RubyModule module = getModuleForAutoload(context, recv);
 
-        if (!IdUtil.isValidConstantName(nonInternedName)) {
-            throw runtime.newNameError("autoload must be constant name", nonInternedName);
+        if (module == null) {
+            return context.nil;
         }
 
-        final RubyString fileString =
-            StringSupport.checkEmbeddedNulls(runtime, RubyFile.get_path(context, file));
-
-        if (fileString.isEmpty()) throw runtime.newArgumentError("empty file name");
-
-        final String baseName = nonInternedName.intern(); // interned, OK for "fast" methods
-        final RubyModule module = getModuleForAutoload(runtime, recv);
-
-        IRubyObject existingValue = module.fetchConstant(baseName);
-        if (existingValue != null && existingValue != RubyObject.UNDEF) return context.nil;
+        return module.autoload_p(context, symbol);
+    }
 
-        module.defineAutoload(baseName, new RubyModule.AutoloadMethod() {
+    @JRubyMethod(required = 2, module = true, visibility = PRIVATE, reads = {CLASS})
+    public static IRubyObject autoload(ThreadContext context, final IRubyObject recv, IRubyObject symbol, IRubyObject file) {
+        final RubyModule module = getModuleForAutoload(context, recv);
 
-            public RubyString getFile() { return fileString; }
+        if (module == null) {
+            throw context.runtime.newTypeError("Can not set autoload on singleton class");
+        }
 
-            public void load(final Ruby runtime) {
-                final String file = getFile().asJavaString();
-                if (runtime.getLoadService().autoloadRequire(file)) {
-                    // Do not finish autoloading by cyclic autoload
-                    module.finishAutoload(baseName);
-                }
-            }
-        });
-        return context.nil;
+        return module.autoload(context, symbol, file);
     }
 
     @Deprecated
     public static IRubyObject autoload(final IRubyObject recv, IRubyObject symbol, IRubyObject file) {
         return autoload(recv.getRuntime().getCurrentContext(), recv, symbol, file);
     }
 
-    static RubyModule getModuleForAutoload(Ruby runtime, IRubyObject recv) {
-        RubyModule module = recv instanceof RubyModule ? (RubyModule) recv : recv.getMetaClass().getRealClass();
-        if (module == runtime.getKernel()) {
-            // special behavior if calling Kernel.autoload directly
-            module = runtime.getObject().getSingletonClass();
+    static RubyModule getModuleForAutoload(ThreadContext context, IRubyObject recv) {
+        RubyModule target = context.getCurrentStaticScope().getModule();
+
+        while (target != null && (target.isSingleton() || target.isIncluded())) {
```

Comment:
@headius So if I autoload in a module which has been included then it migrates to where that was included?  If that module is included in two places then does only the first one fire and get stuck somewhere else?  This easily may be the semantics....

Reply:
Yes that appears to be the semantics. In MRI this logic is rb_class_real, our "getRealClass" but that method lives on RubyClass and not RubyModule.